File: Tree/BinaryTreePostorderTraversal/BinaryTreePostorderTraversal.py
"""
class TreeNode:
    def __init__(self, v):
        this.val = val
        this.left, this.right = None, None 
"""
class Solution:
    def postOrderTraversal(self, root):
        if root is None:
            return []
            
        return self.preOrderTraversal(root.left)  + self.preOrderTraversal(root.right) + [root.val]
        
        
    def postOrderTraversal(self, root, flag):
        if root is None:
            return []
            
        result = []
        s = []
        
        while root or s:
            if root:
                s.append(root)
                root = root.left
            else:
                root = s.pop()
                result.append(root.left)
                root = root.left
            root = s.pop()
                   
        return result         

File: Tree/BinaryTreePostorderTraversal/BinaryTreePostorderTraversal.java
/*
class TreeNode {
	int val;
	TreeNode left, right;
	TreeNode(int v) {
		this.left = null;
		this.right = null;
		this.val = v
	}
}
*/
class Solution {
	public List<Integer> postOrderTraverse(TreeNode root) {
		List<Integer> result = new List<Integer>();
		traverse(root, result);
		return result;
	}
	
	private void traverse(TreeNode root, List<Integer> result) {
		if(root != null) {
			traverse(root.left);
			traverse(root.right);
			result.add(root.val);
		}
	}
	
	public List<Integer> postOrderTraverse(TreeNode root, int flag) {
		List<Integer> result = new ArrayList<Integer>();
		if(root == NULL)
			return result;
		
		Deque<Integer> s = new ArrayDeque<Integer>();
		s.push(root);
		while(root != null && !s.isEmpty()) {
			TreeNode node = s.pop();
			if(root != null) {
				s.push(root);
				root = root.left;
			} else {
				root = s.pop();
				result.add(root.val);
				root = root.right;
			}
			
		}
		return result;
	}
}